File: Competitions/NCPC 2024/AvoidingTheAbyss.py
xs, ys = (int(i) for i in input().split())
xt, yt = (int(i) for i in input().split())
xp, yp = (int(i) for i in input().split())

inf = 1000000

def safe(x, y):
    safe = None
    if y <= yp:
        return (x, -inf), (0, -inf)
    elif x <= xp:
        return (-inf, y), (-inf, 0)
    elif y >= yp:
        return (x, inf), (0, inf)
    elif x >= xp:
        return (inf, y), (inf, 0)

(x1, y1), (x2, y2) = safe(xs, ys)
(x5, y5), (x4, y4) = safe(xt, yt)

points = [(x1, y1), (x2, y2)]

if x2 == x4 and abs(y2 - y4) > inf:
    points.append((inf, 0))
elif y2 == y4 and abs(x2 - x4) > inf:
    points.append((0, inf))
points.extend([(x4, y4), (x5, y5)])

print(len(points))
for x,y in points:
    print(x, y)

File: Kattis/avoidingtheabyss.hs
import           Control.Arrow              ((&&&), (>>>))
import qualified Data.ByteString.Lazy.Char8 as C
import           Data.Function              ((&))
import           Data.Functor               ((<&>))
import           Data.Maybe                 (fromJust)

main :: IO ()
main = do
    s:t:p:_ <- C.getContents <&> (C.lines >>> map (C.words >>> map readInt >>> head &&& last))

    let (a@(x1,y1), b@(x2,y2)) = safe p s
        (e@(x5,y5), d@(x4,y4)) = safe p t
        c | x2 == x4 && abs(y2 - y4) > inf = [(inf, 0)]
          | y2 == y4 && abs(x2 - x4) > inf = [(0, inf)]
          | otherwise                      = []
        points = [a,b] ++ c ++ [d,e]

    points
        & map (\(x,y) -> show x <> " " <> show y)
        & (show (length points) :)
        & map C.pack
        & C.unlines
        & C.putStr

safe :: (Int,Int) -> (Int,Int) -> ((Int,Int), (Int,Int))
safe (px,py) (x,y) | y <= py   = ((x   , -inf), (0    , -inf))
                   | x <= px   = ((-inf, y   ), (-inf , 0   ))
                   | y >= py   = ((x   , inf ), (0    , inf ))
                   | otherwise = ((inf , y   ), (inf  , 0   ))

inf :: Int
inf = 10^7

readInt :: C.ByteString -> Int
readInt = C.readInt >>> fromJust >>> fst
